(BOLT-830) Handle running scripts on windows targets

donoghuc opened the pull request:
Previously running a script on windows targets would not work over the PCP transport. This commit adds logic in the bolt_shim::script task to handle running scripts on windows targets. This logic requires that the script `extension` is sent along with the script `content` and `arguments`. In order to maintain backwards compatability this is optional metadata. An error will be raised if the `extension` metadata is not specified and the target OS is determined to be windows.

donoghuc committed:
(BOLT-830) Handle running scripts on windows targets

Previously running a script on windows targets would not work over the PCP transport. This commit adds logic in the `bolt_shim::script` task to handle running scripts on windows targets. This logic requires that the script file `name` (including extension) is sent along with the script `content` and `arguments`. In order to maintain backwards compatability this is optional metadata. An error will be raised if the task `name` metadata is not specified and the target OS is determined to be windows.

donoghuc committed:
(maint) Pull in updates from pdk 1.8.0

Update the travis configuration and the Gemfile based on PDK 1.8.0

adreyer commented on tasks/script.json:
```diff
@@ -10,6 +10,10 @@
       "type": "Array[String]",
       "description": "Arguments to pass to the script"
+    },
+    "extension": {
```
The other transports seem to create a tmpdir and copy the file with it's original filename. We should probably do that here too.

donoghuc commented:
I have a PE setup with a windows agent if reviewer would like to use that. 
master
```
- t3nch8gvokuhgku.delivery.puppetlabs.net (redhat-7-x86_64-pix, 2.83/48 hours, beaker_version: 4.5.0, department: unknown, project: Beaker, created_by: cas, name: rhel7_master_and_agent, roles: master, agent, database, dashboard, classifier, default)
```
windows agent
```
my284gh0cuifaql.delivery.puppetlabs.net (win-10-ent-x86_64, 2.48/48 hours)
```
linux agent 
```
dwaidedymlw8r3r.delivery.puppetlabs.net (centos-7-x86_64-pix, 2.83/48 hours, beaker_version: 4.5.0, department: unknown, project: Beaker, created_by: cas, name: centos7_agent, roles: agent, frictionless)
```
UPDATE
These have been destroyed.

donoghuc commented on tasks/script.rb:
```diff
@@ -5,25 +5,137 @@
+      ScriptBlock = (Get-Command "#{script_path}").ScriptBlock
+      ArgumentList = @()
+    }
+    #{mapped_args}
```
need to clean that up.

adreyer commented on tasks/script.rb:
```diff
@@ -5,25 +5,137 @@
+      case Pathname(path).extname.downcase
+      when '.rb'
+        [
+          'ruby.exe',
```
Will this be puppets ruby?

adreyer replied:
I'm ok not officially supporting this for now.

Iristyle replied:
Oddly Ruby doesn't have a built-in mechanism for resolving the calling binary. (The only other place Ruby calls `GetModuleFileName` is in its code to dump VMs - https://github.com/ruby/ruby/blob/fc14c408aa0d75693d2e296e5b4a8e7406004aa2/vm_dump.c#L685)

That said... *if* we know this runs on top of the Puppet Ruby interpreter (and we know that we have FFI available as a dependency), this code at https://stackoverflow.com/a/25929455 will work (I would make some minor mods to what's being done here, but the basic premise is correct):

```
[1] pry(main)> require 'ffi'
=> true
[2] pry(main)> module Helpers
[2] pry(main)*   extend FFI::Library
[2] pry(main)*
[2] pry(main)*   ffi_lib :kernel32
[2] pry(main)*
[2] pry(main)*   typedef :uintptr_t, :hmodule
[2] pry(main)*   typedef :ulong, :dword
[2] pry(main)*
[2] pry(main)*   attach_function :GetModuleHandle, :GetModuleHandleA, [:string], :hmodule
[2] pry(main)*   attach_function :GetModuleFileName, :GetModuleFileNameA, [:hmodule, :pointer, :dword], :dword
[2] pry(main)*
[2] pry(main)*   def self.actualRubyExecutable
[2] pry(main)*     processHandle = GetModuleHandle nil
[2] pry(main)*
[2] pry(main)*     # There is a potential issue if the ruby executable path is
[2] pry(main)*     # longer than 999 chars.
[2] pry(main)*     rubyPath = FFI::MemoryPointer.new 1000
[2] pry(main)*     rubyPathSize = GetModuleFileName processHandle, rubyPath, 999
[2] pry(main)*
[2] pry(main)*     rubyPath.read_string rubyPathSize
[2] pry(main)*   end
[2] pry(main)* end
=> :actualRubyExecutable
[3] pry(main)> Helpers.actualRubyExecutable
=> "C:\\tools\\ruby25\\bin\\ruby.exe"
```

I've noticed that there is already a `WinHelpers` class that's being called in this shim - but not sure where that code lives? Might be useful to land this there, depending on how this ships, etc.

donoghuc replied:
For some context, In order to support the local transport on windows bolt runners we pulled some powershell specific code into template functions https://github.com/puppetlabs/bolt/blob/master/lib/bolt/transport/powershell.rb which allowed us to share between the local and winrm transports. My thought was to apply those in this case as well. I chose to keep the bolt_shim::script implementation a single file to support older agents that did not allow multi file tasks.

I tested running the following ruby script on a windows 10 agent:
test.rb:
```
puts ARGV[0]
puts 'hi'
```
Invocation:
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt/Boltdir$ bolt script run test.rb 'im a param' -n hcbig1skqbpx759.delivery.puppetlabs.net --transport pcp
Started on hcbig1skqbpx759.delivery.puppetlabs.net...
Finished on hcbig1skqbpx759.delivery.puppetlabs.net:
  STDOUT:
    im a param
    hi
Successful on 1 node: hcbig1skqbpx759.delivery.puppetlabs.net
Ran on 1 node in 1.45 seconds
```
Feel free to test with them. The only setup I did on master was to clone this repo in `/etc/puppetlabs/code/environments/production/modules/bolt_shim` and checkout this PR. The windows node was simply set up with the `https://r8hrv1pfr6dk15b.delivery.puppetlabs.net:8140/packages/current/install.ps1` script

master:
`r8hrv1pfr6dk15b.delivery.puppetlabs.net`
Windows agent:
`hcbig1skqbpx759.delivery.puppetlabs.net`

Iristyle replied:
Per in-person discussion, we're going to raise an error on executing Ruby files on Windows for now, given there are likely no active users of this functionality.

donoghuc committed:
(BOLT-830) Only support powershell scripts for windows targets

Trying to find the correct interpreter for scripts other than powershell adds complexity to this task and is not likely to be very portable. This commit allows powershell scripts to be run and should satisfy user demand for now. A clean error message will be raised if the target system is windows and a non-powershell script is requested to be run.

donoghuc commented:
Local tests showing behavior:
test.ps1
```
Write-Output 'hi'
Write-Output $args[0]
```
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt/Boltdir$ bolt script run test.ps1 hi -n hcbig1skqbpx759.delivery.puppetlabs.net --transport pcp
Started on hcbig1skqbpx759.delivery.puppetlabs.net...
Finished on hcbig1skqbpx759.delivery.puppetlabs.net:
  STDOUT:
    hi
    hi
Successful on 1 node: hcbig1skqbpx759.delivery.puppetlabs.net
Ran on 1 node in 1.60 seconds
cas@cas-ThinkPad-T460p:~/working_dir/bolt/Boltdir$ bolt script run test.rb hi -n hcbig1skqbpx759.delivery.puppetlabs.net --transport pcp
Started on hcbig1skqbpx759.delivery.puppetlabs.net...
Failed on hcbig1skqbpx759.delivery.puppetlabs.net:
  The command failed with exit code 1
  STDERR:
    Error: Only powershell scripts are supported on windows targets.
Failed on 1 node: hcbig1skqbpx759.delivery.puppetlabs.net
Ran on 1 node in 1.42 seconds
```
